File browser extensions

vitvakatu opened the pull request:
### Pull Request Description

Closes #12004

Note that the extension lists are mocked in the videos below, so you will not see the same result in the app until the libs configuration is updated.

In cloud browser:

https://github.com/user-attachments/assets/be921752-3f75-4ff9-be33-1879f8fef708

In native browser:

https://github.com/user-attachments/assets/066c1e88-9bce-4dc5-8e1c-3973fbbf5948

Latest version (includes changes from the latest develop):

https://github.com/user-attachments/assets/d3c7a6d3-e8ad-4edb-967a-81e34f359f38

In read-mode file browser:

https://github.com/user-attachments/assets/de4cec3b-1f90-4014-aef6-d0931f643de7

### Important Notes

Interaction handler now supports nested interactions. I’m not 100% happy with the API, but I suggest gather more usage patterns before improving it if necessary.

Also, deep refactoring of the SelectionSubmenu widget, making it more universal.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.
- [x] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

vitvakatu committed:
Add file extension text input to file browser widget

vitvakatu committed:
Fix scrolling of file browser widget

vitvakatu committed:
Adding SelectionSubmenu dropdown for file extension input

vitvakatu committed:
Nested interactions

vitvakatu committed:
Filtering file list

vitvakatu committed:
native file browser

vitvakatu committed:
abstract nested selection menu API

vitvakatu committed:
Fix dropdown interaction

vitvakatu committed:
lib-facing API

vitvakatu committed:
fix dynamic config format

vitvakatu committed:
Config-based extensions list

vitvakatu committed:
converting fileetensionentries

vitvakatu committed:
Rename fields of selection submenu entries

vitvakatu committed:
Refactorings and fixes

vitvakatu committed:
Fix typecheck and unit test failure

vitvakatu committed:
Update changelog

farmaazon commented on app/gui/src/project-view/components/ConditionalTeleport.vue:
Seeing "conditional" in the code I would start looking for a condition. But this is not blocking, as I don't have much better name now.

vitvakatu replied:
I have no good answer here… It’s `conditional` because it is conditionally present depending on `target`.

farmaazon commented on app/gui/src/project-view/components/GraphEditor/widgets/WidgetCloudBrowser.vue:
```diff
@@ -24,12 +24,18 @@ const path = computed(() => {
     setActivity(
       computed(() =>
-        h(FileBrowserWidget, {
+        h(FileBrowserWidget as Component, {
```
What changed here, that we need this cast?

vitvakatu replied:
Honestly, not sure, I only added another prop. The typescript breaks in this place without the explicit cast, though, and the error is not even about my added prop: 
<img width="1141" alt="image" src="https://github.com/user-attachments/assets/2e2df76f-eb2c-4d7d-b341-34bcae8b16ad" />

vitvakatu replied:
I think it is caused by `withDefaults` for props of `FileBrowserWidget`, but I’m not sure.

vitvakatu replied:
After seeing it a second time, I believe it is a bug in `withDefaults` that is getting triggered by using optional (`| undefined`) props.

farmaazon commented on app/gui/src/project-view/components/GraphEditor/widgets/WidgetFileBrowser.vue:
```diff
@@ -88,13 +90,39 @@ function makeValue(
+        const extensions = fileType.extensions as string[]
+        return [{ name, extensions }]
+      } else {
+        const nestedFileTypes = fileType.extensions as FileType[]
```
This cast is a bit suspicious. Isn't `FileType[]` the only viable alternative? But perhaps I give typescript checker too much credit.

vitvakatu replied:
Yep, it is as stupid as it is. I remember about type checkers I added in another module though, so I will use them here.

farmaazon commented on app/gui/src/project-view/components/GraphEditor/widgets/WidgetSelection/SelectionSubmenu.vue:
```diff
@@ -104,30 +92,44 @@ defineExpose({
     :show="props.show && submenu != null"
-    :entries="submenuEntries"
-    :selectedExpressions="props.selectedExpressions"
+    :entries="submenuEntries as T[]"
```
This looks strange. The elements are oficially specified as `SubmenuEntry<T>`, why cast them to `T`?

I guess you had a problem with empty array (so typescript make computed value of type `SubmenuEntry | never`) - but here you should just specify type in computedRef, or cast the empty array.

vitvakatu replied:
You are right; I got lost in type conversions somewhere in the middle of the refactoring. The type of submenuEntries is obviously `T[]`, it was just specified incorrectly.

farmaazon commented on app/gui/src/project-view/components/GraphEditor/widgets/WidgetSelection/tags.ts:
```diff
@@ -209,7 +210,7 @@ export interface Actions {
 
 /** A helper type for all possible dropdown entries. */
-export interface Entry extends DropdownEntry {
+export interface Entry extends DropdownEntry, SubmenuEntry<Entry> {
```
SubmenuEntry already extends DropdownEntry, doesn't it?

vitvakatu replied:
Yep. Fixed.

farmaazon commented on app/gui/src/project-view/components/widgets/FileBrowserWidget.vue:
```diff
@@ -298,102 +321,240 @@ onMounted(() => {
+            class="inputField"
+            @click="openDropdown()"
+            @input="fileExtensionInputChanged"
           />
```
The component became a bit overblown. Could we make FileExtensionInput a separate component, and put all related logic there?

farmaazon replied:
Also, in current implementation, we don't have extension picker in read mode. But I'm ok with adding it in a separate PR.

vitvakatu replied:
Kaz did refactoring work, and support for the read mode was added in the last update.

farmaazon commented on app/gui/src/project-view/components/widgets/FileBrowserWidget/__tests__/fileExtensionFilter.test.ts:
Browsing tests, I miss case for accepting with name with/without extension and extension chosen.

vitvakatu replied:
Added these tests now.

farmaazon commented on app/gui/src/project-view/components/widgets/FileBrowserWidget/paths.ts:
```diff
@@ -64,6 +65,11 @@ export function useFileBrowserStack(
   listDirectory: (dir: Directory) => Promise<readonly AnyAsset<AssetType>[]>,
 ) {
   const filenameInputContents = ref<string>('')
+  const fileExtensionInputContents = ref<string>('')
```
The docs of this composable are not updated. Also, either change the name, or split to more composables with well-defined logics.

vitvakatu replied:
Kaz did the refactoring.

farmaazon commented on app/gui/src/project-view/providers/interactionHandler.ts:
```diff
@@ -44,37 +70,70 @@ export class InteractionHandler {
 
-  /** Cancel the current interaction, if it is the specified instance. */
+  /**
+   * Cancel the interaction, if it is currently active.
```
The active is ambiguous, because `isActive` does not consider _parents_ active, but here we do.

vitvakatu replied:
Rephrased

farmaazon commented on app/gui/src/project-view/providers/interactionHandler.ts:
```diff
@@ -44,37 +70,70 @@ export class InteractionHandler {
 
-  /** TODO: Add docs */
-  handleCancel(): boolean {
+  /**
+   * Cancel all interactions, in the order from child to parent.
+   * @returns `true` if the current interaction was cancelled.
+   */
+  cancelAll(): boolean {
```
Why in `cancelAll` we care about the canceling order, but in `cancel` we don't? Same question to `end`.

vitvakatu replied:
We don’t care much now, but I think this property can be helpful in some situations. Made it more consistent.

farmaazon replied:
Only one thing: I imagine a situation, when someone in the future would like to change the order, but feel constrained, assuming that this guarantee is necessary. So I would keep the consistency, but remove the guarantee from docs.

vitvakatu replied:
I think it is not an issue with our current workflow (we frequently refactor the code from the ground up, no matter the complications), but I agree that there is such a risk in principle. I have removed these comments.

farmaazon commented on app/gui/src/project-view/providers/interactionHandler.ts:
```diff
@@ -44,37 +70,70 @@ export class InteractionHandler {
+   * Only the current interaction is considered, its ancestors are not notified.
    */
-  handlePointerEvent<HandlerName extends keyof Interaction>(
+  handlePointerEvent<
+    HandlerName extends keyof Omit<Interaction, 'cancel' | 'end' | 'parentInteraction'>,
+  >(
```
Do we need to omit? I see the non-event-handler interaction fields are checked in line 138.

vitvakatu replied:
Yeah, it is needed, because the check on line 138 was not working correctly (you could set up a handler for `cancel`, for example). Turning `handler` type to `never` does not prevent you from calling the function, unlike the restriction on the generic. Thinking about it now, I will remove all this typescript magic and leave only `'pointerup' | 'pointerdown'` type constraint.

farmaazon commented on app/common/src/fileFilter.ts:
```diff
@@ -0,0 +1,5 @@
+/** Describes allowed file types in the file browser. */
+export interface FileFilter {
+  name: string
+  extensions: 'all' | string[]
+}
```
We could put that in GUI package. The GUI is devDependency of electron client anyway.

vitvakatu replied:
Done

kazcw commented on app/gui/src/project-view/components/ConditionalTeleport.vue:
```diff
@@ -0,0 +1,12 @@
+</script>
+
+<template>
+  <Teleport v-if="props.target" :to="props.target" v-bind="$attrs">
+    <slot />
+  </Teleport>
+  <slot v-else v-bind="$attrs" />
```
What is the purpose of the `v-bind`? It seems hard to use correctly since any properties passed this way must be valid properties to apply to either the `Teleport` or its contents

vitvakatu replied:
Yeah, I guess it is not really needed here. I wanted it for some reason, but forgot why.

kazcw commented on app/gui/src/project-view/components/GraphEditor/widgets/WidgetSelection/SelectionSubmenu.vue:
```diff
@@ -104,30 +92,44 @@ defineExpose({
 <template>
-  <Teleport v-if="props.rootElement" :to="props.rootElement">
-    <div ref="dropdownElement" :style="floatingStyles" class="SelectionSubmenu widgetOutOfLayout">
+  <ConditionalTeleport :target="props.rootElement" v-bind="$attrs">
```
`defineOptions({ inheritAttrs: false })` is needed to prevent properties from being passed twice

vitvakatu replied:
I’m not sure this is strictly needed, but adding it wouldn’t hurt. I think it is not required because the component already has multiple roots.

kazcw commented on app/gui/src/project-view/components/GraphEditor/widgets/WidgetSelection/SelectionSubmenu.vue:
```diff
@@ -104,30 +92,44 @@ defineExpose({
 <template>
-  <Teleport v-if="props.rootElement" :to="props.rootElement">
-    <div ref="dropdownElement" :style="floatingStyles" class="SelectionSubmenu widgetOutOfLayout">
+  <ConditionalTeleport :target="props.rootElement" v-bind="$attrs">
```
All the contents of the tree of `SelectionSubmenu` seem to ultimately be teleported to the root's `props.rootElement`. Does this teleportation need to be part of the `SelectionSubmenu`, or could we just put the root `SelectionSubmenu` in a `Teleport`?

vitvakatu replied:
Teleport is required for node dropdowns to work (I don’t remember the exact specifics, but nodes are heavily broken without it). So, moving teleport to the outside of the component is not generally useful — we still need to use the same teleport target for nested submenus, so we would still pass an additional prop.

kazcw commented on app/gui/src/project-view/components/widgets/FileBrowserWidget.vue:
```diff
@@ -26,15 +37,22 @@ import Backend, {
+    writeMode: false,
+    choosenPath: '',
+    type: 'file',
+    fileTypes: () => [{ label: 'All files', extensions: ['*'] }],
```
Why the lambda?

vitvakatu replied:
It is how `withDefaults` works, see an example in the docs: https://vuejs.org/api/sfc-script-setup.html#default-props-values-when-using-type-declaration

vitvakatu committed:
Fixing behavior of the file extension filter

kazcw commented on app/gui/src/project-view/providers/interactionHandler.ts:
```diff
@@ -26,15 +29,38 @@ export class InteractionHandler {
+    active: WatchSource<boolean>,
+    interactionBuilder: (parent: Interaction | undefined) => Interaction,
+  ) {
+    const activeInteraction = ref<Interaction>()
```
I think an `onCleanup` in the `watch` could make `activeInteraction` unnecessary.

vitvakatu replied:
Yep, fixed

vitvakatu committed:
Move FileFilter type to enso-gui

vitvakatu committed:
Addressing minor review comments

vitvakatu committed:
simplify interaction handler implementation

vitvakatu committed:
remove unneeded fallthrough attributes

vitvakatu committed:
wip: Move file name bar to a separate component

vitvakatu committed:
Merge with develop

vitvakatu committed:
Merge with develop

vitvakatu committed:
Fixing remaining conflicts, integrating changes

vitvakatu committed:
Close file browser on accept

vitvakatu committed:
Avoid the need in nextTick when updating the contents of the input
fields

vitvakatu committed:
Correct modal display, fix style for acceptance button

vitvakatu committed:
Show file extension picker in read-mode file browser, fix empty
configuration case.

vitvakatu committed:
Remove comments about interactions handling order

vitvakatu commented:
Merge with develop done, addressed final review comments, and added support for read-only file browser mode. I also fixed a few spotted bugs in the develop code:
- Z-ordering of the modal was broken, allowing interaction with “hidden” buttons
- CSS for the confirmation button was broken
- The widget was not closing after accepting a file, which I assume is not an expected behavior (compared to ~3 weeks old develop)

Updated the top post with additional videos, ready for re-review.

vitvakatu committed:
regenerate engine docs

farmaazon commented on app/gui/src/project-view/components/widgets/FileBrowserWidget/FileBrowserNameBar.vue:
```diff
@@ -1,18 +1,159 @@
+  setFilter: [Filter]
 }>()
+
+// === File Extension Filter ===
+
```
It's ok to go, but I would still at some point make it a component separate from FileBrowserNameBar - just not not have everything prefixed with `fileExtension` ;)

vitvakatu replied:
Ugh, I tried doing that, but splitting the components is difficult because they are interconnected. I will leave it as it is.

vitvakatu committed:
Merge with develop

vitvakatu commented:
@AdRiley @jdunkerley could you review the libs modifications?

vitvakatu committed:
Merge with develop
